Network call using async task

chelseybaker opened the pull request:
[Trello Card](https://trello.com/c/x6JJAftp/8-re-implement-network-call-using-async-task)

This builds on #8 and #7 

This doesn't need to be merged (the Trello card says it's basically just to use AsyncTask and can be discarded). It can just be closed (but I would prefer to save the branch). 

AsyncTask is nice, I don't know what that guy was talking about :P

stkent commented on app/src/main/java/networking/TenDayAsyncResponse.java:
```diff
@@ -0,0 +1,11 @@
+ * Created by chelseybaker on 10/20/16.
+ */
+
+public interface TenDayAsyncResponse {
```
A more typically Androidy name for this interface would be `OnTenDayForecastDownloadedListener`. (I'm not afraid of long names :P). See e.g. the classic `View.OnClickListener`, whose method is then named `onClick`. Here that'd mean we'd use a method name like `onTenDayForecastDownloaded`.

stkent commented on app/src/main/java/networking/AsyncTaskRunner.java:
```diff
@@ -0,0 +1,53 @@
+
+public class AsyncTaskRunner extends AsyncTask<Call<TenDayForecast>, Void, TenDayForecast> {
+
+    public TenDayAsyncResponse delegate = null;
```
In Java we basically _never_ use public instance variables like this. Instead, make this `private final` and initialize it in the task constructor (since there's no point having a task instance where the response is ignored, in this case).

stkent replied:
The `delegate` name is rarely used in Android; prefer `listener` or `callback` depending on the interface name.

stkent commented on app/src/main/java/networking/AsyncTaskRunner.java:
```diff
@@ -0,0 +1,53 @@
+ * Created by chelseybaker on 10/20/16.
+ */
+
+public class AsyncTaskRunner extends AsyncTask<Call<TenDayForecast>, Void, TenDayForecast> {
```
Similar to how we name `Activity`/`Fragment` subclasses, I'd suggest it's more typical to leave `AsyncTask` as a suffix, e.g. `TenDayForecastAsyncTask`.

stkent commented on app/src/main/java/com/detroitlabs/chelseybaker/wundergroundweather/activities/LandingActivity.java:
```diff
@@ -160,29 +183,9 @@ private void advanceToForecastWithLocation(String location, final Boolean listVi
-            }
-        });
+        AsyncTaskRunner asyncTaskRunner = new AsyncTaskRunner();
+        asyncTaskRunner.delegate = this;
```
See other comment r.e. injecting this via the constructor. If constructor injection wasn't appropriate, you'd still want a private instance field and a public setter instead of a public instance field.

stkent commented on app/src/main/java/com/detroitlabs/chelseybaker/wundergroundweather/activities/LandingActivity.java:
```diff
@@ -19,15 +19,19 @@
+        GoogleApiClient.ConnectionCallbacks, TenDayAsyncResponse {
+
+    // Holds whether the user selected to view the forecast in a list view or page view
+    private Boolean isListView = false;
```
From the other PR, but: should be safe to use the primitive `boolean` type here. We only need the boxed types `Integer`, `Float`, etc. in a few situations:

- if used in a collection type, e.g. `List<Integer>` (`List<int>` is not allowed);
- if the value can be null (`Boolean` values can be null, but `boolean` values cannot and default to `false`).

stkent commented:
You're right, `AsyncTask` isn't bad. It is, however, pretty much totally unaware of Android lifecycle (with respect to configuration changes, like orientation changes, especially) so it's really only recommended for short-running asynchronous jobs. From the docs:

> AsyncTasks should ideally be used for short operations (a few seconds at the most.)

I'd never use it for networking now, since we have such good libraries available, but it's good to know it exists. For example, in Herbie, we use an `AsyncTask` to assemble a layered vehicle image.

stkent commented:
@chelseybaker comments in!

chelseybaker commented:
I renamed and made notes according to the naming styles. I also changed `LandingActivity.java` to use Retrofit's callback style again, but kept the AsyncTask code in there for reference

stkent commented on app/src/main/java/com/detroitlabs/chelseybaker/wundergroundweather/activities/LandingActivity.java:
```diff
@@ -146,10 +146,10 @@ public void onConnectionSuspended(int i) {
 
     @Override
-    public void forecastProcessFinished(TenDayForecast tenDayForecast) {
+    public void onTenDayForecastDownloaded(TenDayForecast tenDayForecast) {
```
❤️

stkent commented on app/src/main/java/networking/TenDayForecastAsyncTask.java:
```diff
@@ -16,9 +16,13 @@
-    public TenDayAsyncResponse delegate = null;
+    private final OnTenDayForecastDownloadedListener delegate;
+
+    public TenDayForecastAsyncTask(OnTenDayForecastDownloadedListener delegate) {
```
Could optionally annotate this parameter as `@NonNull` to warn consumers you are expecting a non-null instance (you dereference it on line 55 without a null check). Silly Java NPEs!

stkent commented:
Cool beans, couple minor notes added. Should I merge this or close it?

chelseybaker commented:
They have been marked @NonNull!

stkent commented:
Woot, perfect! Asking again since I'm not sure - do you want this merged or closed? I believe the Trello card indicated closing was ok but you made a few retrofit tweaks too so...?
